### ISSUE TITLE
distributed resizing

### ISSUE THREAD

**Issue by pohly** [kubernetes-csi/external-resizer#142]
For local storage (e.g. LVM disks) it makes sense to support a deployment model where the driver only runs locally on nodes together with sidecars. This is already support for provisioning (kubernetes-csi/external-provisioner#524, https://github.com/kubernetes-csi/external-provisioner#deployment-on-each-node).

**Comment by pohly** [comment]
In the resizer, the sidecar probably needs to check whether the volume is local before invoking the CSI driver. This can be achieved via the existing topology field. https://github.com/kubernetes-csi/external-provisioner/issues/592 might help to reduce traffic by configuring the informer so that it filters by label and thus only receives PVs that are of interest. See https://github.com/kubernetes-csi/external-provisioner/pull/583 for code which does such server-side filtering for CSIStorageCapacity objects.

**Comment by travisghansen** [comment]
How much effort is required for this?

**Comment by pohly** [comment]
I suspect it'll be easier than distributed provisioning because the sidecar can already determine from the PV whether it is responsible for the operation. See https://github.com/kubernetes-csi/external-snapshotter/pull/585 for the corresponding feature in the snapshotter.

**Comment by travisghansen** [comment]
A fair point.

My sense is that controllers generally have no knowledge or understanding of how the device is being used and is just expanding raw capacity while node operations (when necessary) are meant to deal with the specifics of ensuring the volume (generally a block device if node resizing is involved) capacity is readily accessible to the workload (generally resizing the fs as necessary). I prefer to keep these logical separations.

With that in mind, I have several common code paths that are used by both distributed and centralized deployments (my project currently has 17 distinct 'drivers') so I'd love to be able to keep those logically in the same buckets.

From a broader (than k8s) perspective, I think the 'deploy the controller alongside the node' workflow is clearly a goal of csi so, regardless of any perceived value/arguments/points I make above, making assumptions about what makes sense or not probably doesn't serve the interests of the spec in the long-term. The spec is interested in facilitating node+controller deployments jointly and part of controller deployments is resizing...by extension csi compliant COs should be capable of invoking resize operations on controllers deployed in such a fashion.

In short, IMO regardless of specific merits for my use-case, it seems a supported csi deployment style that should 'just work' with k8s and/or any other COs.

**Comment by gnufied** [comment]
> The spec is interested in facilitating node+controller deployments jointly and part of controller deployments is resizing...by extension csi compliant COs should be capable of invoking resize operations on controllers deployed in such a fashion.

The spec does not say anything about node+controller deployment on same nodes. I am still unsure, what problems distributed resizing will solve that is not already solved by calling `NodeExpandVolume` on the node.

**Comment by travisghansen** [comment]
> The spec does not say anything about node+controller deployment on same nodes. I am still unsure, what problems distributed resizing will solve that is not already solved by calling `NodeExpandVolume` on the node.

Isn't that figure 2 and 3 here?: https://github.com/container-storage-interface/spec/blob/master/spec.md

Again, regardless of that, IMO the provisioner and snapshotter both support node+controller deployments. The fact that k8s has various clients/code-bases to invoke the grpc methods is an implementation details. If k8s supports node+controller it should support it fully and not selectively decide which grpc methods it will or will not invoke based on supposition about the merits/requirements of the feature by csi driver implementations from specific SPs.

**Comment by pohly** [comment]
@gnufied: if the CSI driver only runs on the nodes, where would you run external-resizer?

If you only run one instance of it per cluster, then it has no CSI driver instance that it can call. Or is the expectation that a CSI driver then has an implementation that runs on the node and another implementation that runs alongside the external-resizer, with the second implementation not doing any actual work?

**Comment by gnufied** [comment]
Currently - if a CSI driver supports node-only volume expansion, the external-resizer does provide a no-op trivial resizer which is used for updating PV (kubelet can't update PVs) and rest of the volume expansion happens on kubelet via `NodeExpandVolume`.

**Comment by pohly** [comment]
I agree that this seems to be sufficient for node-local storage. Is this documented sufficiently so that CSI driver developers can find out about it?

**Comment by gnufied** [comment]
I get the point about deploying controller side of the plugin on the node and not having any explicit control-plane associated with CSI driver. I need to think some more about it. I have scheduled this item for discussion on Monday's CSI implementation meeting. If you can manage, please join so as we can talk about it.

**Comment by pohly** [comment]
> I don’t think it’s appropriate to selectively decide which controller methods will/will not be invoked in node+controller setups based on supposition.

My understanding is that the controller side will only be skipped if a CSI driver developer explicitly asks for it by deploying the CSI driver in a suitable way. That seems okay to me. How that works specifically I don't know (hence my question about documentation).

**Comment by travisghansen** [comment]
OK thanks for pointing out the doc!

IMO this furthers my points about the spec. Stating the perhaps obvious, the sidecar (to my knowledge) must be deployed as..a sidecar. It cannot be deployed standalone without the/a controller deployed along with it. That deployment story does not match any of the architectures from the spec. It's a sort of mashing together of 1+2 and/or 1+3 where 'node' hosts are running controller+node plugins and then a 'master' host is running a controller plugin. Or said differently, 'controller' plugin is running both on 'nodes' and 'master' hosts.

**Comment by zimnx** [comment]
> This sidecar is needed even if CSI driver does not have EXPAND_VOLUME controller capability, in this case it performs a NO-OP expansion and updates PersistentVolume object.

In case when there's an error during expanding, shouldn't this cause update to fail? 
Controller deployed as deployment, can't check whether there's enough space to expand on different node.
Leaving PV in expanded state even though expansion didn't happen is wrong in my view.

CSI Spec also [requires specific errors](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume-errors) to be returned when Controller isn't able to expand the volume, so no-op expansion is not in sync with the spec.

**Comment by mnencia** [comment]
The core argument @pohly raised earlier in this thread was never answered: in a node-only DaemonSet deployment (no central controller pod), there is nowhere to run a centralized external-resizer. The provisioner and the snapshotter both support `--node-deployment` for exactly this deployment model. The resizer is the only sidecar that doesn't, which makes it impossible to offer volume expansion in node-only setups where the driver implements `ControllerExpandVolume`.

A workaround exists for drivers that can disable controller expansion and rely solely on `NodeExpandVolume` (called by kubelet locally). But that forces driver authors to give up controller-side expansion logic to accommodate a gap in the resizer, which shouldn't be necessary when the other sidecars already support this deployment model.

I'm exploring the possibility of enhancing the csi-driver-host-path distributed deployment to use `--node-deployment` for all sidecars (kubernetes-csi/csi-driver-host-path#651), which would also serve as an integration test for this feature. The previous PR #195 was closed by the triage bot. I'm happy to open a fresh PR rebased on current master if there's interest.

**Comment by travisghansen** [comment]
Yes please :) A problem with node expand is it doesn’t include the same information that controller expand does (ie: size iirc).

I believe we should just support this model as you have mentioned.
